x-pack/plugins/apm/server/routes/__test__/routeFailures.test.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Server } from 'hapi';
+// @ts-ignore
+import { initErrorsApi } from '../errors';
+import { initServicesApi } from '../services';
+// @ts-ignore
+import { initStatusApi } from '../status_check';
+import { initTracesApi } from '../traces';
+import { initTransactionsApi } from '../transactions';
+
+async function testRouteFailures(init: (server: Server) => void) {
+  const mockServer = { route: jest.fn() };
+  init((mockServer as unknown) as Server);
+  expect(mockServer.route).toHaveBeenCalled();
+
+  const routes = mockServer.route.mock.calls;
+  const mockReq = {
+    params: {},
+    query: {},
+    pre: {
+      setup: {
+        config: { get: jest.fn() },
+        client: jest.fn(() => Promise.reject(new Error('request failed')))
+      }
+    }
+  };
+
+  routes.forEach(async (route, i) => {
+    test(`route ${i + 1} of ${
+      routes.length
+    } should fail with a Boom error`, async () => {
+      try {
+        await route[0].handler(mockReq);
+      } catch (err) {
+        expect(err.isBoom).toEqual(true);
+        return;
+      }
+
+      throw new Error('Route succeeded when it should have failed');
+    });
+  });
+}
+
+describe('route handlers fail properly', () => {
+  let consoleError: () => void;
+
+  beforeEach(() => {
+    consoleError = window.console.error;
+    window.console.error = jest.fn();
```

Comment:
Shouldn't this be `global` instead of `window` (there is no `window` object in Node)

Reply:
Afaik the mock will be cleaned up automatically if you use `jest.spyOn`:

```js
jest.spyOn(global.console, 'error')
```

Reply:
Hm good point, although it works 🤔 probably because jsdom env running in jest?

Reply:
ah, that might be. Can you check if `global` works too? That seems more "right".

Reply:
I'm not sure if spies can have their return values changed? But even if they can, the damage console.error does is a side effect and not its return value, so this won't help -- the method has to be replaced. 

I'll switch it to `global` though if that still works! :+1:

Reply:
`undefined` _is_ the return value of `console.log()`. It's just a shorthand for this:
```js
jest.spyOn(global.console, 'error').mockImplementation(() => undefined)
```

Reply:
Right, but `console.error` already returns `undefined` all on its own right? It's the side effect of printing to stdout that I'm trying to suppress, and changing its return value won't do that.

Reply:
> Right, but console.error already returns undefined all on its own right? It's the side effect of printing to stdout that I'm trying to suppress, and changing its return value won't do that.

Both of these will supress the side effect:

```js
jest.spyOn(global.console, 'error').mockImplementation(() => undefined)
 # and
jest.spyOn(global.console, 'error').mockReturnValue(undefined);
```

Simply doing this won't:
```js
jest.spyOn(global.console, 'error')
```

Reply:
Ok looked into this and TIL you can _stub_ a function in jest using spyOn ... did not expect mockImplementation that to work like that!